platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/schemaBuilder/DataSchemaBuilder.kt

```diff
@@ -204,7 +205,32 @@ class DataSchemaBuilder(
         }
 
         private fun isAllowedTypeParameter(typeParameter: KTypeParameter): Boolean =
-            currentContextStack.any { contextElement -> contextElement is SchemaBuildingContextElement.ModelMemberContextElement && contextElement.kCallable.typeParameters.any { it == typeParameter } }
+            currentContextStack.any { contextElement ->
+                contextElement is SchemaBuildingContextElement.ModelMemberContextElement && contextElement.kCallable.typeParameters.any {
+                    typeParameterMatches(it, typeParameter)
+                }
+            }
+
+        /**
+         * Workaround: the Kotlin standard library functions have their type parameter `T` represented by different [KTypeParameter] objects when seen in the functions [KFunction.typeParameters] and when
```

Comment:
Maybe it would be worth putting this into `platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/workarounds.kt` so that we might remember it and replace it in the future.

Reply:
Done in [`365f734` (#32409)](https://github.com/gradle/gradle/pull/32409/commits/365f7343c63122ca1d5ae5fe88d6c7d1c14c4baa)